[DRAFT] Engine plugin API and engine entry point for Lloyd's KMeans

ogrisel opened the pull request:
This is a draft pull-request to allow third-party packages such as https://github.com/soda-inria/sklearn-numba-dpex to contribute alternative implementations of core computational routines of CPU-intensive scikit-learn estimators.

This would be particularly useful to experiment with GPU-optimized alternatives to our CPU-optimized Cython code.

This PR will serve as a design experiment to tackle #22438.

ogrisel committed:
wip engines

ogrisel committed:
wip

ogrisel committed:
fixes

ogrisel committed:
wip

ogrisel committed:
more specific assertion

ogrisel committed:
Add docstring to the config context

ogrisel committed:
add default kwarg

ogrisel committed:
Various fixes

ogrisel committed:
empty doc

ogrisel committed:
Merge branch 'main' into wip-engines

ogrisel committed:
WIP

ogrisel committed:
Merge branch 'main' into wip-engines

ogrisel committed:
wip

ogrisel committed:
Move tolerance computation to the engine

ogrisel committed:
wip

ogrisel committed:
Merge branch 'main' into wip-engines

ogrisel committed:
wip

ogrisel committed:
wip

ogrisel commented:
/cc @fcharras @jjerphan @betatim

ogrisel commented:
Note: this is still a work in progress and the Engine API is not set in stones. In particular, I broken the MBKMeans tests. But this should be enough to get started.

ogrisel commented on sklearn/cluster/_kmeans.py:
```diff
@@ -1407,66 +1507,28 @@ def fit(self, X, y=None, sample_weight=None):
-                verbose=self.verbose,
-                tol=self._tol,
-                n_threads=self._n_threads,
             )
```
Note that currently we delegate the full Lloyd loop to the engine for the sake of simplicity for this first iteration.

However my plan would rather be to only delegate one Lloyd iteration at a time instead and move the loop and convergence check back into the estimator class (probably in a new private method).

Moving this loop back into the estimators class will be required to properly integrate @jeremiedbb's work on callbacks for instance: #22000.

betatim replied:
While looking at the `KMeans` in cuml and sklearn I noticed that my test data set that took ~2.5s for sklearn took only 25ms in cuml. I assume with numba-dpex the speed up will be similar. So the main point is "something that takes long now might be super fast in a plugin".

On the one hand `max_iter` is typically(?) only O(100) so the overhead of making a few hundred additional function calls is not that great. On the other hand, what data would the callbacks need to make decisions (beyond just progressbar updates, which you could argue you don't need for fast stuff) and how much does it cost to transfer that from device to device?

My question is: do you/someone know what happens to the performance of alternative implementations of kmeans when they have to be implemented as "one iteration at a time"? I don't know enough about how kmeans on something like a GPU works, but I hope the answer is "nothing happens, this is totally doable" because that would make life easy.

fcharras replied:
In general it will be implemented "one interation at a time" (exemple with the plugin we're working on: https://github.com/soda-inria/sklearn-numba-dpex/blob/main/sklearn_numba_dpex/kmeans/drivers.py#L342, or likewise with the current cython implementation in sklearn: https://github.com/scikit-learn/scikit-learn/blob/17df37aee/sklearn/cluster/_kmeans.py#L647) and the loop on `max_iter` is in python, it is completely negligible.

ogrisel replied:
I expect that that for most algorithms in scikit-learn, one fit iteration would last 10x or more the overhead of a GPU dispatch (assuming we leave the data and temporary datastructures on device memory between consecutive iterations).

betatim replied:
What is the plan regarding this change? I think it would make sense to already go for the "one call per iteration" (what you propose) now instead of waiting. If that works for you I'll send a PR.

betatim replied:
It will also help sharpen the API for when data is passed in to an engine. My proposal would to settle on:

* `prepare_fit(X, y, sample_weight)` will be called every time when `KMeans.fit()` is called. The engine should keep hold of the data as it won't be provided again
* the single iteration method becomes `kmeans_single(current_centroids)`. `X` and friends isn't passed again.

In particular when we call `kmeans_single` once per iteration we need a way to provide the data to the engine once, so that it can convert it or decide that it would like to raise `NotImplemented`, etc. If `X` and friends are passed to `kmeans_single` on each iteration, the engine should in principle be prepared to convert it at each iteration because there is no promise that it hasn't change since the last time `kmeans_single` was called.

Something that would be harder to do in that case is re-use things from the Cython engine. For example currently I use:
```python
def init_centroids(self, X):
    return cp.asarray(super().init_centroids(X))
```
because there is not much benefit in re-implementing the initial centroid selection. In a super strict world where the above proposal is adopted, then `init_centroids` should also not be passed `X` again, but it should use what was previously provided in `prepare_fit()`.

The point being, I'd support switching to the per-iteration API so we can work these things out with code (instead of our heads).

fcharras replied:
I think we agree on the implementation strategy here so you're welcome to send a PR  :+1: and let's iterate on the details from there if needed, I think @ogrisel @jjerphan and I can be available for review.

Regarding:

> because there is not much benefit in re-implementing the initial centroid selection.

Actually I can think of two reasons:

- initial centroid selection includes k-means++ which might benefit from optimizations

- it's not likely that 'super().init_centroids(X)` will work if `X` is not a numpy or scipy array, and it could be a good thing for the `fit` to accept other type of inputs when used with other engines, e.g `cupy` arrays for a a cuda-based engine, to save data loading to and from the targeted device. So, in 'super().init_centroids(X)` might need to be rewritten for cases where `X` is a different object.

fcharras committed:
linting

fcharras commented on sklearn/_config.py:
```diff
@@ -14,6 +14,7 @@
     ),
     "enable_cython_pairwise_dist": True,
     "array_api_dispatch": False,
+    "engine_provider": (),
```
to enforce the experimental aspect of this api, we could have an additonal switch "`use_pluggable_engines`"  `False` by default that disable the `engine_provider` keyword if not forcibly set to `True` ?

fcharras committed:
fix MBKMeans and linting

ogrisel committed:
Merge branch 'main' into wip-engines

ogrisel committed:
Draft changelog entry

ogrisel committed:
doc reorg

ogrisel committed:
fix changelog entry to add the pr number

ogrisel committed:
fix test name

ogrisel committed:
attempt at sphinx the sphinx warning

fcharras committed:
fix MBKMeans test

fcharras committed:
fix: skip entry points that do not match the requested provider names (rather than the opposite)

fcharras committed:
update config_context unit test to account for new engine_provider keyword

fcharras committed:
for python < 3.10 returns an empty list when the slearn engines entry point is missing (rather than erroring out))

ogrisel committed:
Link to user guide from docstring

fcharras committed:
add a verbosity parameter to get_engine_class

betatim commented:
A thought on naming: should we call the engines "engines" or something else like "plugin"? The reason I was thinking about this is the discussion at yesterday's meeting about tree splitters. It made me think that maybe the use case of quantile forest, oblique trees, etc is another example where having a plugin mechanism would be a good thing. In this case you'd not be replacing a computational bit of code like for kmeans, but providing additional splitters/criteria or hooking other parts of the algorithm.

It is more of a naming nitpick, but also naming is important :-/

fcharras committed:
add pytest plugin that can be used by engine providers to run sklearn tests with customised engines

fcharras committed:
Add plugin methods for predict, transform and score methods for KMeans

fcharras committed:
ad _engine.base module

fcharras committed:
linting

fcharras committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into wip-engines

ogrisel commented:
> A thought on naming: should we call the engines "engines" or something else like "plugin"? 

The scope of this engine PR is only provide alternative implementations for low level components of existing estimators. It is not meant to extend the functional scope and public API of scikit-learn in any way.

I think it's perfectly fine to provide new estimators in independent libraries (no need for an extension point system in that context). If they want to claim scikit-learn compatibility they can just use the `sklearn.utils.estimator_checks.check_estimator` tool as documented in https://scikit-learn.org/stable/developers/develop.html#rolling-your-own-estimator.

fcharras committed:
fix a bug that caused by the pytest plugin where sklearn tests would be executed twice

fcharras committed:
__all__fixup and renaming FeatureNotCoveredByPluginError -> NotSupportedByEngineError

fcharras committed:
exception __all__ fixup

fcharras commented:
I struggled a bit for setting up the pytest plugin, it finally works (with the help of `pytest.hookimpl`) except for windows but I don't have a mean of reproducing windows errors at the moment, does the issue is familiar to you ? (basically pytest failing to load the plugin entry point, saying the module where it's defined is not found..)

fcharras committed:
register the _engine subpackage in setup.py

fcharras committed:
add __init__.py

fcharras commented:
> pytest failing to load the plugin entry point

fixed in 70e4ecfdc9dcc57622339d095028e764c484c674

fcharras committed:
fix test_engines test

fcharras commented:
@jeremiedbb this is the test for which the success depends on how `np.argpartition` breaks ties:

https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/tests/test_k_means.py#L89

The points `(0, 0)` and `(1,1)` are equally distant from the centroid `(0.5, 0.5)` so both could be seen as valid relocations ?

The backend we're working on ties the other way and the test fail, not sure what we should do there.

betatim commented:
Some thoughts that came up while building a plugin (sorry for the unsorted listed):
* when is the right moment for the plugin to raise `NotImplemented`? Maybe in the `prepare_fit` call, I think it has access to all the information it needs to make a decision
* what is a good way for engines to re-use (parts of) the default implementation? Is that a sign that the engine methods should be smaller?
* when does a plugin engine get imported? I'm seeing that the first `fit()` takes a lot longer than the second, third, n-th call. This gives plugin a disadvantage when benchmarking (when people don't pay attention).
* when should the engine store state? Similar to estimators that have to handle all state in `fit()` and none in `__init__`? Maybe this isn't important to worry about, but I'm not sure.
* it is a bit messy to implement only the fit part or only the predict/transform part. In my case after implementing the `fit()` part attributes on the estimator end up being `cupy` arrays which means I also have to implement `predict()`.
* related to the previous item: what should the type of estimator attributes be? I can see the case for wanting them to be numpy arrays, no matter what plugin you used. I can also see the case for the being cupy arrays if you used a plugin that uses cupy. For the "attributes as numpy" I end up storing them also as cupy (to avoid having to constantly re-create them), which links back to the question about "where should engines store state".
* testing. Making sure an estimator with/without a plugin turned on gives "the same" answers is a bit tricky. You can't just `np.allclose` because the order of centroids might be different, and then you need to work out which labels are equivalent ,etc. Not rocket science but also more than 5minutes of work. Might be good to work out some "centralised" methods for making this easier

fcharras commented:
It's awesome to hear that you're also making progress on plugin investigations. Maybe you have a link to share for the plugin you're working on ?

For some of the points you raise I think we have come up with practical implementations that are rather sensible:

- for the testing part I would suggest seeing the small [pytest extension](https://github.com/scikit-learn/scikit-learn/pull/24497/files#diff-c2637b5713dfb63e1aa63af34dd6ea7aaffcb6d2455bb5da4dec6cdc0d723b4aR8) we add in this branch and how we use it in the test bench for the gpu kmeans [here](https://github.com/soda-inria/sklearn-numba-dpex/blob/main/.github/workflows/run_tests.yaml#L35), if also features the use of a special exception for cases where no implementation is available for the set of input parameters. We've ran into issues regarding not being able to fully conform to all sklearn tests because of design differences in the very engine that cannot be overcome, like random generation or how to break ties in sorting algorithm, in the current state the engines can't pass those tests and must port them. Maybe the design in sklearn can be improved in this regard, by adding special flags to categorize engine-specific tests so it can be disabled and/or ported easily by plugin providers, and try to not have engine-specific tests where it's not necessary to begin with. 

- for questions about interoperability between engines, we thought that adding an interface to convert fitted estimators from and to engines would be a good feature, much like it has been done in the [array API PR](https://github.com/scikit-learn/scikit-learn/pull/22554)

- I think the others points you raise are implementation details that should be more discussed at the engine levels since sklearn can't enforce how much code is shared between the vanilla engine and an external engine. About fixed overhead when benchmarking, we start a [dry run](https://github.com/soda-inria/sklearn-numba-dpex/blob/main/benchmark/kmeans.py#L105) before running the actual benchmark.

betatim commented:
A super simple implementation of an alternative engine with GPU support is in https://github.com/betatim/scikit-learn-gpu

betatim commented:
Another to do item: if we keep the estimator attributes in the "native" format of the engine, then users get weird error messages if they accidentally have code like:

```python
with sklearn.config_context(engine_provider="sklearn_gpu"):
  km = KMeans()
  km.fit(X)

# Then some time later, maybe in a different file, with a different human
km.predict(new_X)
```
I think providing a friendly error message to the user would be nice. Maybe the way to do this is to store the name of the engine that was used during `fit()` and inform the user that they are using a different one at `predict()` time (and that the engine provides a function for converting?)?

fcharras committed:
fix _config.py conflicts

fcharras commented:
I pulled master to include https://github.com/scikit-learn/scikit-learn/pull/24779 in this branch

Some negative about directly using entry points for registering engines:

> with the current state of our plugin system users can't just write a new engine in a jupyter notebook and load it into the plugin api, it seems that entry points are required to be declared in the setup file.

betatim commented:
> @ogrisel, @fcharras, and @betatim: I think it would be valuable that we discuss this PR synchronously during our call tomorrow to sync and follow up. In that case, we must report notes of our discussions in a comment here so that readers and reviewers have all the informations and rationales for any decisions.
> 
> What do you think of this proposal?

Yes please. In particular I'm interested in what people think of #24826 vs having a bespoke API per estimator.

ogrisel commented:
Unfortunately I am running late on other things and I won't be able to attend the meeting tomorrow. But I will try to catchup with the PR later this week.

jjerphan committed:
Merge branch 'main' into wip-engines

jjerphan committed:
fixup! Merge branch 'main' into wip-engines

jjerphan commented on doc/computing/engine.rst:
```diff
@@ -0,0 +1,29 @@
+Computation Engines (experimental)
+==================================
+
+**This API is experiment** which means that it is subject to change without
+any backward compatibility guarantees.
```
```suggestion
**This API is experimental** which means that it is subject to change without
any backward compatibility guarantees.
```

jjerphan commented on sklearn/cluster/tests/test_bicluster.py:
```diff
@@ -253,7 +253,6 @@ def test_spectralbiclustering_parameter_validation(params, type_err, err_msg):
 
 @pytest.mark.parametrize("est", (SpectralBiclustering(), SpectralCoclustering()))
 def test_n_features_in_(est):
-
```
This is a spurious change which I think should be reverted.

jjerphan commented on sklearn/exceptions.py:
```diff
@@ -38,6 +39,19 @@ class NotFittedError(ValueError, AttributeError):
 
+class NotSupportedByEngineError(NotImplementedError):
+    """External plugins might not support all the combinations of parameters and
+    input types that the the vanilla sklearn implementation otherwise supports. In such
```
Thanks for adding the purpose and the specification of failing tests' semantic.
```suggestion
    input types that the vanilla sklearn implementation otherwise supports. In such
```

fcharras replied:
note that @betatim also have a proposal for this, relying on an `accept` method exposed by engines rather than catching an exception.

jjerphan commented on sklearn/exceptions.py:
```diff
@@ -38,6 +39,19 @@ class NotFittedError(ValueError, AttributeError):
+class NotSupportedByEngineError(NotImplementedError):
+    """External plugins might not support all the combinations of parameters and
+    input types that the the vanilla sklearn implementation otherwise supports. In such
+    cases, plugins can raise this exception class. When running the sklearn test modules
```
```suggestion
    cases, plugins can raise this exception. When running the sklearn test modules
```

jjerphan commented on doc/whats_new/v1.2.rst:
```diff
@@ -2,6 +2,26 @@
+
+TODO: move to doc/whats_new/v1.3.rst once it exists in main:
+
+- |Enhancement| Experimential engine API (no backward compatibility guarantees)
```
```suggestion
- |Enhancement| Experimental engine API (no backward compatibility guarantees)
```

jjerphan commented on sklearn/_engine/base.py:
```diff
@@ -0,0 +1,112 @@
+    for provider_name in provider_names:
+        spec = specs_by_provider.get(provider_name)
+        if spec is not None:
+            # XXX: should we return an instance or the class itself?
+            return spec.get_engine_class()
```
Based on the current name of the function, we must return the class.

Do want want caller to handle objects' creation using the class or not? If yes, returning the class is more appropriate.

jjerphan commented on sklearn/cluster/_kmeans.py:
```diff
@@ -256,15 +257,148 @@ def _kmeans_plusplus(X, n_clusters, x_squared_norms, random_state, n_local_trial
+        if tol == 0:
+            return 0
+        if sp.issparse(X):
+            variances = mean_variance_axis(X, axis=0)[1]
```
I find this is a bit more understandable: it makes explicit that `means` and `sum_weights` are ignored. 
```suggestion
            _, variances, _ = mean_variance_axis(X, axis=0)
```

What do you think?

jjerphan commented on sklearn/cluster/_kmeans.py:
```diff
@@ -1866,6 +2067,15 @@ def __init__(
 
+        if self.tol > 0:
+            if sp.issparse(X):
+                variances = mean_variance_axis(X, axis=0)[1]
```
Ditto: I find this is a bit more understandable: it makes explicit that `means` and `sum_weights` are ignored. 
```suggestion
            _, variances, _ = mean_variance_axis(X, axis=0)
```

What do you think?

jjerphan reviewed (commented):
Here is a first pass, I have not yet had enough time to build a proper understanding of the proposals.

fcharras committed:
add engine method to get nb of distinct clusters

fcharras committed:
Merge branch 'wip-engines' of https://github.com/ogrisel/scikit-learn into wip-engines

fcharras committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into wip-engines

fcharras commented on sklearn/cluster/tests/test_bicluster.py:
```diff
@@ -253,7 +253,6 @@ def test_spectralbiclustering_parameter_validation(params, type_err, err_msg):
 
 @pytest.mark.parametrize("est", (SpectralBiclustering(), SpectralCoclustering()))
 def test_n_features_in_(est):
```
```suggestion
def test_n_features_in_(est):

```

fcharras committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

fcharras committed:
linting

fcharras committed:
fix

fcharras commented:
WDYT of https://github.com/soda-inria/sklearn-numba-dpex/pull/65#issuecomment-1327620854 ?

fcharras committed:
fix

betatim committed:
Merge branch 'main' into wip-engines

betatim committed:
Add support for trying engines in turn

Users can specify a list of engine providers to try, this change makes
it so we try them in order until one is found that accepts the request.

betatim committed:
Fix formatting

betatim committed:
Update mean_variance computation

betatim commented:
Updated this to include the "try different providers in turn" functionality.

For now I stuck with using an explicit `accepts()` function. This seemed simpler to implement for now compared to having a "input validation and acceptance decision" method that might raise an exception. Another thought I had is that the "should I accept this or not?" decision should be cheap to call, so that we can try several engine providers in turn. If they each do expensive computations to then reject the request that seems like a waste. And if the acceptance computation is cheap, it matters less if it is repeated in the input validation (if it needs repeating and can't be factored out somehow, given that engine instances are short lived).

Also made a small tweak to make the kmeans tests pass again.

ogrisel committed:
Move changelog entry to v1.3.rst

ogrisel committed:
Merge main

ogrisel commented:
I merged main and moved the changelog entry to target 1.3.

fcharras commented:
I adapted the plugin in https://github.com/soda-inria/sklearn-numba-dpex/pull/74/files

betatim committed:
Pass sample_weight parameter

Co-authored-by: Franck Charras <29153872+fcharras@users.noreply.github.com>

betatim committed:
Add engine aware mixin to factor out engine stuff

betatim committed:
Add attribute conversion decorator

betatim committed:
Tweak attribute conversion

betatim committed:
Use Array API to get unique cluster count

betatim committed:
Rename cluster counting method

betatim committed:
Update conversion related bits

betatim committed:
Fix engine provider at fit time

betatim committed:
Combine engine selection and validation

betatim committed:
Rename argument

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel committed:
Merge pull request #14 from betatim/engines-cluster-count

Rename cluster counting method

jjerphan committed:
Merge branch 'main' into wip-engines

betatim committed:
Rename loop variable

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

betatim committed:
Switch back to using `accepts` for engine selection

betatim committed:
Allow "runtime" engines to be passed as well as provider names

betatim committed:
Update tests

betatim committed:
Update docstring for `accepts`

betatim committed:
Update comment for engine class config

betatim committed:
Add `engine_name` attribute to ad-hoc engine classes

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

betatim committed:
Merge pull request #13 from betatim/engine-mixin

fcharras commented:
Last round of merges broke the tests :red_circle: apparently some issue with the engine getter raising `RuntimeError`s

ogrisel commented:
@fcharras suggested that the last merge broke the CI without us realizing it because https://github.com/ogrisel/scikit-learn/pull/13 targeted the `ogrisel/scikit-learn` repo instead of `scikit-learn/scikit-learn` repo and as a results the regular CI config does not apply.

I think I should push the `wip-engines` branch on `ogrisel/scikit-learn` to a new long-running feature branch (e.g. named `feature-engine-api`) on the main `scikit-learn/scikit-learn` repo. This way, sub-PRs to this new feature branch would benefit from the usual CI setup without the extra maintenance of reconfiguring CI servies to work for the `ogrisel/scikit-learn` repo.

Any opinion or better name name suggestion @scikit-learn/core-devs?

/cc @betatim

ogrisel commented:
When doing so I will probably squash the existing history as most of the original commits have non-informative wip commit messages. Let me know if you are opposed to this plan. Shared co-authorship of the squashed commit should be preserved.

jjerphan commented:
This looks appropriate to me. 👍

Do you think `feature/engine-api` is slightly clearer?

betatim commented:
Sounds like a good idea and it will be nice to have the benefit of robots checking our work :)

No opinion on the branch name.

ogrisel commented:
Closing in favor of the newly created #24497. Let's stop using the `ogrisel/scikit-learn:wip-engines` branch from now on.
